### DISCUSSION TITLE
How to run snapshot tests on docfx forked repository

### DISCUSSION THREAD

**Discussion by filzrev** [dotnet/docfx#9743] [Show and tell]
When running snapshot tests on local environment with docfx forked repository.
A lot of diffs occurs.

It can be reduced by following steps.

**1. Enable snapshot tests**
By default. SnapShot tests is skipped for non-CI environment.
It need to set environment variable to run these tests.

1.1. Edit`test\docfx.Snapshot.Tests\ModuleInitializer.cs`
1.2. Add following line.> Environment.SetEnvironmentVariable("SNAPSHOT_TEST", "EnableSnapshotTest");

**2. Modify `src\Docfx.Common\Git\GitUtility.cs` file**

Modify following functions to replace repository dependent string.
- `TryGetFileDetail`
- `RawContentUrlToContentUrl`
- `GetSourceUrl`

(e.g. `url.Replace("filzrev", "dotnet").Replace(".git", "");

**Comment by yufeih** [comment]
OS/hardware setup may affect screenshot test output, so it is excluded from locally. We can perhaps replace the current approach with a visual regression test service such as percy which is free to open source projects. Other suggestions on visual regression services are also welcome.

**Comment by yufeih** [comment]
We are now using Percy for visual testing and snapshot tests can run locally now.

**Comment by filzrev** [comment]
Thanks for updates.
Is it expected to create my own `percy` account to tests?

I'm trying to run percy tests with following steps.
And successfully compare screenshots.

**Steps to execute percy tests at local environment**

1. Create `https://percy.io` account. And create project.
2. Install percy CLI. with following command 
 > npm i -g @percy/cli
3. Create snapshots of `main` branch

```powershell
git switch main
$env:PERCY_TOKEN = '[Write-only Token]'

dotnet build -c Release
percy exec -- dotnet test -c Release -f net8.0 --no-build --filter Stage=Percy
```

4. Approve `main` branch snapshots from percy's project web site.

5. Create snapshots of `my-feature-branch`.

```powershell
git switch [my-feature-branch]
$env:PERCY_TOKEN = '[Write-only Token]'
$env:PERCY_TARGET_BRANCH = 'main'

dotnet build -c Release
percy exec -- dotnet test -c Release -f net8.0 --no-build --filter Stage=Percy
```

**Reply by yufeih** [reply]
No - there were sth wrong with PERCY_TOKEN not exposed to forked PRs which should be resolved now.
